### PR TITLE
GTFS FLEX Support - `core-utils` changes only

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
This PR contains all `core-utils` changes needed for #297.

### CODE REVIEW

File: packages/core-utils/package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
   "private": false,
   "dependencies": {
     "@mapbox/polyline": "^1.1.0",
-    "@opentripplanner/geocoder": "^1.1.0",
+    "@opentripplanner/geocoder": "^1.1.1",
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/package.json:16]: Latest is now 1.1.2.

File: packages/core-utils/src/itinerary.js [hunk 2]
```diff
@@ -26,6 +26,23 @@ export function isTransit(mode) {
   return transitModes.includes(mode) || mode === "TRANSIT";
 }
 
+export function isOptional(leg) {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/itinerary.js:29]: Rename to `isReservationRequired` or `isBookingRequired` (to match the return condition).

File: packages/core-utils/src/itinerary.js [hunk 3]
```diff
@@ -26,6 +26,23 @@ export function isTransit(mode) {
...
+  return leg.boardRule === "mustPhone";
+}
+export function isContinuousDropoff(leg) {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/itinerary.js:32]: Add JSDoc explaining why only "Must coordinate with driver" is used and not other continuous dropoff values.

File: packages/core-utils/src/itinerary.js [hunk 4]
```diff
@@ -26,6 +26,23 @@ export function isTransit(mode) {
...
+  return leg.alightRule === "coordinateWithDriver";
+}
+export function isFlex(leg) {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/itinerary.js:35]: Add JSDoc to explain why we use the leg boarding/alight rules mentioned and not others.

File: packages/core-utils/src/itinerary.js [hunk 5]
```diff
@@ -26,6 +26,23 @@ export function isTransit(mode) {
...
+  return isOptional(leg) || isContinuousDropoff(leg);
+}
+
+export function bookingInfoDropofHasBookingTime(info) {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/itinerary.js:39]: Rename to `isAdvanceBookingRequired`.

File: packages/core-utils/src/itinerary.js [hunk 6]
```diff
@@ -26,6 +26,23 @@ export function isTransit(mode) {
...
+  return info?.latestBookingTime?.daysPrior > 0;
+}
+export function legDropoffHasBookingTime(leg) {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/itinerary.js:42]: Maybe rename to `legDropoffRequiresAdvanceBooking` ?

File: packages/core-utils/src/query.js [hunk 7]
```diff
@@ -163,13 +163,32 @@ function isParamApplicable(paramInfo, query, config) {
   return true;
 }
 
+/**
+ * Helper method which replaces OTP flex modes with single FLEX mode that's
+ * more useful and easier to work with.
+ */
+export function reduceOtpFlexModes(modes) {
+  return modes.reduce((prev, cur) => {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query.js:171]: It looks like [JavaScript's `Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) would fit well here (but still return an array).

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
See the renaming and documentation comments, and other suggestions.

File: packages/core-utils/src/query.js (hunk 7 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/query.js:171]: I'm a little confused as to how you see Set helping. The source array doesn't contain duplicates, the issue is taking `FLEX_DIRECT`, `FLEX_ACCESS`, and `FLEX_EGRESS` and combining those into the single `FLEX` item. If you have a cleaner solution I would love it!

File: packages/core-utils/package.json (hunk 1 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/package.json]: upgraded in adf39ed79fe9ef36db930bb0c93555320077dc92

File: packages/core-utils/src/itinerary.js (hunk 2 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.js]: Renamed in f7aec8908db009cc5d1b7ddbfe2e062539692f51

File: packages/core-utils/src/itinerary.js (hunk 3 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.js:42]: Added in f7aec8908db009cc5d1b7ddbfe2e062539692f51

File: packages/core-utils/src/itinerary.js (hunk 4 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.js:49]: Added in f7aec8908db009cc5d1b7ddbfe2e062539692f51

File: packages/core-utils/src/itinerary.js (hunk 5 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.js]: Renamed in f7aec8908db009cc5d1b7ddbfe2e062539692f51

File: packages/core-utils/src/itinerary.js (hunk 6 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.js]: I love it! Changed in f7aec8908db009cc5d1b7ddbfe2e062539692f51

File: packages/core-utils/src/query.js [hunk 8]
```diff
@@ -163,13 +163,32 @@ function isParamApplicable(paramInfo, query, config) {
...
+    const newModes = prev;
+    // Add the current mode if it is not a flex mode
+    if (!cur.includes("FLEX")) {
+      newModes.push(cur);
+      // If it is a flex mode, do not add it but rather add the custom flex mode
+      // if not already present
+    } else if (!newModes.includes("FLEX")) {
+      newModes.push("FLEX");
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query.js:179]: Sorry, these were the lines that made me think of `Set`, but that's just a code refinement.

**Review by binh-dam-ibigroup — APPROVED** [review]
Changes look good as proposed.

**Comment by daniel-heppner-ibigroup** [comment]
Don't know entirely what core-utils does, but the code looks good!

**Review by daniel-heppner-ibigroup — APPROVED** [review]

**Pull request merged**